### DISCUSSION TITLE
[Feature Request] Vim-like Navigation in Scrollback & Option to Open Output in a New Tab

### DISCUSSION THREAD

**Discussion by 0x4a61636f62** [ghostty-org/ghostty#3708] [Feature Requests, Ideas] [answered]
In terminals like Alacritty and WezTerm, you can freely navigate the visible text, make selections, and copy/paste using standard Vim keybindings (e.g., `j`, `k`, `v`, `y`, `p`, etc.).

Currently, Ghostty does not offer this. However, users can use `write_scrollback_file:open` or `write_screen_file:open` to open the scrollback in an external editor. While this is less convenient, it provides access to a fully featured editor, complete with navigation features like line numbers, etc.

**Proposed Solution**
Getting the best of both approaches might be challenging. Still, one potential improvement could be allowing the user to open the temporary file in a new tab within Ghostty (via configuration). This could combine some of the benefits of both methods in a user-friendly way that does not create unnecessary technical complexity early in the Ghostty project.

**Comment by ScriptAlchemist** [comment]
I'm not sure it even needs to open into a new tab. In reality if the current tab/panel/window turned into an editor. That could config easily with $EDITOR. If it turned into an editor with no line numbers and no active indicators. You wouldn't know the difference of it going into a copy mode.

It's in theory, the same as going into a copy mode. On something like tmux or screen. Allowing you to move around and copy what you need. Which allows for your own person customization per editor. But having a default to something like vi or nano over TextEditor. Where it stays in the same terminal surface it's called on. Cause popping up a completely new window isn't a nice flow for someone who wants to stay in the terminal.


You can currently do this with `nvim (keybinding to write_scrollback_file:paste)` and then going into the file. But that's just one step too much. IMHO. It should be the standard

**Reply by xzbdmw** [reply]
Open in nvim loses colors, that's the only downside for me...

**Reply by ScriptAlchemist** [reply]
> Open in nvim loses colors, that's the only downside for me...

Yes, that's a problem. But I'm thinking I'll just use tmux for the time being. Until ghostty can fix the scrollback defaults.

**Comment by 0x4a61636f62** [comment]
I agree with you, @ScriptAlchemist, that it is one step to many to have to switch to a new window or tab. The ideal solution, as you rightly say, would be to use the same terminal surface and have the niceties of a fully featured editor of choice inside it, although I don't know how complex that would be to implement.

From what I was told, the main challenge is that Ghostty currently avoids "stealing" input from the shell like Alacritty and WezTerm does to implement this feature. Implementing this feature similarly would probably require hooking specific keybindings in a scrollback context and something tells me that it is not ideal.

One possible approach might be to integrate a lightweight "scrollback editor" that runs inside the same Ghostty surface or open the scrollback in a TUI-based editor but reuse the same Ghostty instance instead of launching an entirely new window. But, I know too little to know if any of this is technically possible.

And, I agree with you that any solution should remain optional and customizable so users who prefer default keybindings or rely on shell vi modes are not disrupted.

**Comment by ScriptAlchemist** [comment]
For example. I just put a solution into my .bashrc and now I probably won't think about this until the update for now

my solution to add to to a .bashrc for the time being.

but it runs the commands and adds it directly into the terminal input and presses enter for me. Which just opens up my vi instance with the returned path from ghostty. Which activates with ctrl + a + [ like my tmux would move to copy mode.

```bash
run_vi_with_filepath() {
  osascript -e 'tell application "System Events" to keystroke "vi + "'
  osascript -e 'tell application "System Events" to keystroke "j" using {command down, option down, shift down}' # keybind = super+alt+shift+j=write_screen_file:paste
  osascript -e 'tell application "System Events" to key code 36' # Key code 36 is 'Enter' on macOS
}

bind -x '"\C-a[":run_vi_with_filepath'

```

I updated this with a commend on what keybinding to add to the ghostty config. To make the keybinding use the write_screen_file because it seems to be working better than the scroll history command at the moment.

**Reply by xzbdmw** [reply]
fish version
```fish
function run_vi_with_filepath
    set fish_bind_mode insert
    # # Tell System Events to type `nvim -c "norm! G"_dk` to delete the last prompt
    osascript -e 'tell application "System Events" to keystroke "nvim -c \'norm! G\"_dd\' "'
    # Press Cmd+Shift+J
    osascript -e 'tell application "System Events" to keystroke "j" using {command down, shift down}'
    # Press Enter (key code 36 on macOS)
    osascript -e 'tell application "System Events" to key code 36'
end

function fish_user_key_bindings
    bind --preset -M insert \x1c run_vi_with_filepath
    # this uses <c-\>
    bind --mode default \x1c run_vi_with_filepath
end
```

**Reply by SilasMarvin** [reply]
For what its worth, if you want a solution that uses only Ghostty I use:
```
keybind = "ctrl+shift+h=write_screen_file:paste"
keybind = "ctrl+shift+j=text:\x01hx +1000000 \n"
```

This is a two step sequence: `ctrl+shift+h` -> `ctrl+shift+j`. 

`write_screen_file:paste` writes the path to the screen and `\x01hx +100000 \n` performs the following sequence of keystrokes:
1. `ctrl+a` to move to the beginning of the line
2. `hx +100000 \n` to insert and execute the command to open helix and go to line 1000000 (which should always be the bottom)

The only downside is that it involves two key commands, but I find it works much faster than using `osascript` and is cross platform.

**Reply by xzbdmw** [reply]
@SilasMarvin Nice solution, here is a karabiner entry to only have one keymap
```json
{
                        "description": "Map Control+\\ to Command+Shift+H and Command+Shift+J",
                        "manipulators": [
                            {
                                "type": "basic",
                                "from": {
                                    "key_code": "backslash",
                                    "modifiers": {
                                        "mandatory": [
                                            "control"
                                        ]
                                    }
                                },
                                "to": [
                                    {
                                        "key_code": "h",
                                        "modifiers": [
                                            "left_command",
                                            "left_shift"
                                        ]
                                    },
                                    {
                                        "key_code": "j",
                                        "modifiers": [
                                            "left_command",
                                            "left_shift"
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
```

**Reply by maximilianigl** [reply]
How would you adapt this to when you're sshed into a server?

**Comment by MithicSpirit** [comment]
I think that the “best of both worlds” would be to just have both options available. While opening the scrollback in an editor can be nice, I've found that editors aren't great at dealing with files with 100k+ lines (which I often get to when compiling stuff), in which case having a built-in modal keyboard-oriented scrollback navigation and selection mode would be nice. This would also combo well with #189, by allowing searching within the navigation mode.

**Reply by ScriptAlchemist** [reply]
I agree with this more than my original statement.

 It's needed overall and it seems like it's already in the works with the select all and the ability to move the selections as well.

**Reply by 0x4a61636f62** [reply]
I also agree and I have seen the issue before. We need a convenient way to navigate the scrollback with vim keys also.

**Reply by MithicSpirit** [reply]
> We need a convenient way to navigate the scrollback with vim keys also.

Yeah, that's basically what Alacritty does, and it's really nice imo (it also allows selecting+copying text and opening links iirc). I'm currently on Kitty due to ligatures, and I use the “pipe scrollback buffer into a program” thing as a stopgap, but I miss Alacritty every time I do so. Definitely switching to Ghostty as soon as this is added.

**Reply by 0x4a61636f62** [reply]
I use it all the time. If we want to improve on Alacrity, it would be to add more Vim features, such as a column with line numbers, syntax highlighting, etc. although I have no idea how complex that would be.

**Comment by dwolrdcojp** [comment]
I would really love to see this added to ghostty. Vim movements in the terminal is the main reason I switched to Alacritty.

**Comment by 00-kat** [comment]
Related: #3488.

**Comment by xzbdmw** [comment]
I'm just opening a full screen nvim builtin terminal for now to move around freely, you can hardly see the difference tbh...

**Reply by 0x4a61636f62** [reply]
I do that, too, as a workaround for now, but it is a different workflow, and I have noticed that the terminal in Neovim behaves slightly differently for me in some regards.

**Reply by pmouraguedes** [reply]
Me too: `nvim -c "terminal" -c "wincmd o"'` and ditto regarding the workflow ^^

**Comment by swaingotnochill** [comment]
I wanted this feature so bad. I have been using this for a long time with kitty and now with ghostty.
I have the below in my .zshrc config. It opens up neovim terminal as a whole. 

```
alias nterm='nvim -c "terminal" -c "setlocal nonumber norelativenumber"'
if [[ -z "$NVIM_TERMINAL" ]]; then
	export NVIM_TERMINAL=1
	nterm
fi
```

**Comment by mitchellh** [comment, marked as answer]
Vim-mode navigation is now mostly possible in `main` with the latest key tables work. Here is the config:

```
# Vim-mode keybindings for Ghostty.
#
# Note: We're missing a number of actions to produce a better vim experience.
# This is what is possible today! But we plan on adding more actions to make
# this even better in the future.

# Entry point
keybind = alt+v=activate_key_table:vim

# Key table definition
keybind = vim/

# Line movement
keybind = vim/j=scroll_page_lines:1
keybind = vim/k=scroll_page_lines:-1

# Page movement
keybind = vim/ctrl+d=scroll_page_down
keybind = vim/ctrl+u=scroll_page_up
keybind = vim/ctrl+f=scroll_page_down
keybind = vim/ctrl+b=scroll_page_up
keybind = vim/shift+j=scroll_page_down
keybind = vim/shift+k=scroll_page_up

# Jump to top/bottom
keybind = vim/g>g=scroll_to_top
keybind = vim/shift+g=scroll_to_bottom

# Search (if you want vim-style search entry)
keybind = vim/slash=start_search
keybind = vim/n=navigate_search:next

# Copy mode / selection
# Note we're missing a lot of actions here to make this more full featured.
keybind = vim/v=copy_to_clipboard
keybind = vim/y=copy_to_clipboard

# Command Palette
keybind = vim/shift+semicolon=toggle_command_palette

# Exit
keybind = vim/escape=deactivate_key_table
keybind = vim/q=deactivate_key_table
keybind = vim/i=deactivate_key_table

# Catch unbound keys
keybind = vim/catch_all=ignore
```

We're going to keep adding more actions to support more features and also probably ship with this table built-in (but unbound) so users can just add `activat_key_table:vim` in their configs to get it.

**Reply by 0x4a61636f62** [reply]
Thank you @mitchellh :pray:

**Reply by mitchellh** [reply]
I want to repeat that there are missing keybind actions to make the vim mode more full featured. Namely, we are missing selection cursors and the ability to start a selection from the keyboard. But importantly, we now must **simply implement generic binding actions** and it composes nicely into the vim mode. 

But, we already can have great navigation! 

I plan on tackling selection cursors soon and looking into other features to continue to make this more fluid.

**Reply by cau1k** [reply]
The future is near 🔮

**Reply by ScriptAlchemist** [reply]
Thank you. This makes my use of the app so much better. Not everyone likes keyboard focused nav, but it's a lifestyle choice I'm happy you're accomidating.

**Reply by pmouraguedes** [reply]
Thank you! I'm currently waiting for this to switch to ghostty :)

**Reply by Guiwoo** [reply]
Awesome 👏

**Reply by stevenxxiu** [reply]
Please stop the +1 comments. *GitHub* has reactions for a reason. There's many people subscribed to this discussion.

**Reply by binc4t** [reply]
1. use the nightly builds from the repo's releases page instead of the stable version from the official website.
2. If you are using `alt+v` on macOS, make sure to set `macos-option-as-alt = true` in your config. Otherwise, you’ll end up typing the √ symbol.

**Reply by davinkevin** [reply]
fyi; I tried it with ZelliJ, and they might conflict a bit so only a sub-set of commands has an effect, unfortunately 😔

**Comment by ScriptAlchemist** [comment]
Does anyone know if they shipped the vim table built in if you're just using the macos downloaded version. I'm on the 1.3.1 and I was hoping to activate this but I'm a little lost on where to find it. Or if I should just use the one that was pasted above.

**Reply by ilyaZar** [reply]
same question here

**Reply by ScriptAlchemist** [reply]
> same question here

@ilyaZar I think the best bet is going to be adding it yourself. I still find it very limiting and hard to really act like a tmux copy mode or anything.

So personally, I'm still opening them up a tmux session myself.

**Reply by ilyaZar** [reply]
@ScriptAlchemist thanks ! sry missed your reply, it worked after adding manually actually as you described
